edgelet/edgelet-http-workload/src/server/cert/mod.rs

```diff
@@ -80,69 +98,125 @@ fn refresh_cert(
+                                    let context_copy = context.clone();
+
+                                    let pk = new_cert_privkey
+                                        .private_key_to_pem_pkcs8()
+                                        .context(context.clone())?;
+                                    let cert = Certificate::new(cert, pk);
+                                    let cert = cert_to_response(&cert, context_copy.clone())?;
+                                    let body = match serde_json::to_string(&cert) {
+                                        Ok(body) => body,
+                                        Err(err) => {
+                                            return Err(Error::from(err.context(context_copy)))
+                                        }
+                                    };
+
+                                    let response = Response::builder()
+                                        .status(StatusCode::CREATED)
+                                        .header(CONTENT_TYPE, "application/json")
+                                        .header(CONTENT_LENGTH, body.len().to_string().as_str())
+                                        .body(body.into())
+                                        .context(context_copy)?;
+
+                                    Ok(response)
+                                });
+                            Ok(response)
+                        })
+                })
+        })
         .flatten()
         .or_else(|e| Ok(e.into_response()));
 
     Box::new(response)
 }
 
-fn generate_key_and_csr(
-    props: &CertificateProperties,
-) -> std::result::Result<(openssl::pkey::PKey<openssl::pkey::Private>, Vec<u8>), ErrorStack> {
+fn generate_local_keypair() -> std::result::Result<
+    (
+        openssl::pkey::PKey<openssl::pkey::Private>,
+        openssl::pkey::PKey<openssl::pkey::Public>,
+    ),
+    ErrorStack,
+> {
     let rsa = openssl::rsa::Rsa::generate(2048)?;
     let privkey = openssl::pkey::PKey::from_rsa(rsa)?;
     let pubkey = privkey.public_key_to_pem()?;
     let pubkey: openssl::pkey::PKey<openssl::pkey::Public> =
         openssl::pkey::PKey::public_key_from_pem(&pubkey)?;
 
+    Ok((privkey, pubkey))
+}
+
+fn load_keypair(
+    ca_key_pair_handle: &aziot_key_common::KeyHandle,
+    key_engine: &mut openssl2::FunctionalEngine,
+) -> std::result::Result<
+    (
+        openssl::pkey::PKey<openssl::pkey::Private>,
+        openssl::pkey::PKey<openssl::pkey::Public>,
+    ),
+    ErrorStack,
+> {
+    let (workload_ca_private_key, workload_ca_public_key) = {
+        let workload_ca_key_pair_handle =
+            std::ffi::CString::new(ca_key_pair_handle.0.clone()).unwrap();
+        let workload_ca_public_key = key_engine
+            .load_public_key(&workload_ca_key_pair_handle)
+            .unwrap();
+        let workload_ca_private_key = key_engine
+            .load_private_key(&workload_ca_key_pair_handle)
+            .unwrap();
```

Comment:
Unless they provably won't panic, these `.unwrap()`s should probably be replaced with some proper error handling, right?

Reply:
Fixed